ssl/priority_queue.c

```diff
@@ -0,0 +1,395 @@
+ * order pre-condition.
+ */
+static ossl_inline void pqueue_move_down(OSSL_PQUEUE *pq, size_t n)
+{
+    struct pq_heap_st *h = pq->heap;
+
+    ASSERT_USED(pq, n);
+    while (n > 0) {
+        const size_t p = (n - 1) / 2;
+
+        ASSERT_USED(pq, p);
+        if (pq->compare(h[n].data, h[p].data) >= 0)
+            break;
+        pqueue_swap_elem(pq, n, p);
+        n = p;
+    }
+}
+
+/*
+ * Move an element up to its correct position to restore the partial
+ * order pre-condition.
+ */
+static ossl_inline void pqueue_move_up(OSSL_PQUEUE *pq, size_t n)
+{
+    struct pq_heap_st *h = pq->heap;
+    size_t p = n * 2 + 1;
+
+    ASSERT_USED(pq, n);
+    if (pq->htop > p + 1) {
+        ASSERT_USED(pq, p);
+        ASSERT_USED(pq, p + 1);
+        if (pq->compare(h[p].data, h[p + 1].data) > 0)
+            p++;
+    }
+    while (pq->htop > p && pq->compare(h[p].data, h[n].data) < 0) {
+        ASSERT_USED(pq, p);
+        pqueue_swap_elem(pq, n, p);
+        n = p;
+        p = n * 2 + 1;
+        if (pq->htop > p + 1) {
+            ASSERT_USED(pq, p + 1);
+            if (pq->compare(h[p].data, h[p + 1].data) > 0)
+                p++;
+        }
+    }
+}
+
+int ossl_pqueue_push(OSSL_PQUEUE *pq, void *data, size_t *elem)
+{
+    size_t n, m;
+
+    if (!ossl_pqueue_reserve(pq, 1))
+        return 0;
+
+    n = pq->htop++;
+    m = pq->freelist;
+    pq->freelist = pq->elements[m].posn;
+
+    pq->heap[n].data = data;
+    pq->heap[n].index = m;
+
+    pq->elements[m].posn = n;
+#ifndef NDEBUG
+    pq->elements[m].used = 1;
+#endif
+    pqueue_move_down(pq, n);
+    if (elem != NULL)
+        *elem = m;
+    return 1;
+}
+
+void *ossl_pqueue_peek(const OSSL_PQUEUE *pq)
+{
+    if (pq->htop > 0) {
+        ASSERT_USED(pq, 0);
+        return pq->heap->data;
+    }
+    return NULL;
+}
+
+void *ossl_pqueue_pop(OSSL_PQUEUE *pq)
+{
+    void *res;
+    size_t elem;
+
+    if (pq == NULL) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+    if (pq->htop == 0) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        return NULL;
+    }
+
+    ASSERT_USED(pq, 0);
+    res = pq->heap->data;
+    elem = pq->heap->index;
+
+    if (--pq->htop != 0) {
+        pqueue_move_elem(pq, pq->htop, 0);
+        pqueue_move_up(pq, 0);
+    }
+
+    pq->elements[elem].posn = pq->freelist;
+    pq->freelist = elem;
+#ifndef NDEBUG
+    pq->elements[elem].used = 0;
+#endif
+    return res;
+}
+
+void *ossl_pqueue_remove(OSSL_PQUEUE *pq, size_t elem)
+{
+    size_t n;
+
+    if (pq == NULL) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+    if (elem >= pq->hmax) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        return NULL;
+    }
+    if (pq->htop == 0) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
+    ASSERT_ELEM_USED(pq, elem);
```

Comment:
If we are doing ERR_R_INTERNAL_ERROR return for empty pq, should we also unconditionally check here and return with the same error if the element is not used?

Reply:
These checks are only meaningful in debug mode.  They introduce an overhead.

It would be quite reasonable to not raise errors for the empty check above or even not have the checks at all.  All of this is completely internal and we don't mandate parameter checking for internal APIs.  I tend to code defensively and include them but they are easy to remove.

Reply:
If I understand the ASSERT_ checks correctly, they are just internal consistence checks mostly which means they are OK to keep as asserts. But this particular place is different - if an external caller calls ossl_pqueue_remove() with elem that is not in the pqueue, it will result in assert in debug build and UB in non-debug build. IMO that is wrong and the public (rest of the library) facing API calls should tolerate calls where pop/remove is called on empty pq or a pq with missing item to be removed.

I'd probably even drop the ERR_R_INTERNAL_ERROR call in these cases.

Reply:
Currently, the information isn't maintained in non-debug tools to make this check.  It could be included but it seems like a fair bit of overhead for one check.

I suspect hitting an unused item is less likely than passing the wrong queue in :grin:

There is one internal error I think should be kept (the return check from compute_pqueue_growth), this won't be fired by a bug in the caller.

I'm almost tempted to remove all the argument and sanity run time checks rather than just the errors.

Reply:
I would probably just document it as a precondition in the manpage that the element must be in the queue, and that undefined behaviour results if it isn't.

Reply:
Good idea, I've add: _It is undefined what happens if an attempt to remove an element that isn't in the queue is made._